capsules/src/tickv.rs

```diff
@@ -0,0 +1,500 @@
+        Self {
+            flash,
+            flash_read_buffer: TakeCell::new(flash_read_buffer),
+            region_offset,
+        }
+    }
+}
+
+impl<'a, F: Flash> tickv::flash_controller::FlashController<512> for TickFSFlastCtrl<'a, F> {
+    fn read_region(
+        &self,
+        region_number: usize,
+        _offset: usize,
+        _buf: &mut [u8; 512],
+    ) -> Result<(), tickv::error_codes::ErrorCode> {
+        if self
+            .flash
+            .read_page(
+                self.region_offset + region_number,
+                self.flash_read_buffer.take().unwrap(),
+            )
+            .is_err()
+        {
+            Err(tickv::error_codes::ErrorCode::ReadFail)
+        } else {
+            Err(tickv::error_codes::ErrorCode::ReadNotReady(region_number))
+        }
+    }
+
+    fn write(&self, address: usize, buf: &[u8]) -> Result<(), tickv::error_codes::ErrorCode> {
+        let data_buf = self.flash_read_buffer.take().unwrap();
+
+        for (i, d) in buf.iter().enumerate() {
+            data_buf.as_mut()[i + (address % 512)] = *d;
+        }
+
+        if self
+            .flash
+            .write_page((0x20040000 + address) / 512, data_buf)
+            .is_err()
+        {
+            return Err(tickv::error_codes::ErrorCode::WriteFail);
+        }
+
+        Err(tickv::error_codes::ErrorCode::WriteNotReady(address))
+    }
+
+    fn erase_region(&self, region_number: usize) -> Result<(), tickv::error_codes::ErrorCode> {
+        self.flash.erase_page(self.region_offset + region_number);
+
+        Err(tickv::error_codes::ErrorCode::EraseNotReady(region_number))
+    }
+}
+
+pub type TicKVKeyType = [u8; 8];
+
+pub struct TicKVStore<'a, F: Flash + 'static> {
+    tickv: AsyncTicKV<'a, TickFSFlastCtrl<'a, F>, 512>,
+    operation: Cell<Operation>,
+    next_operation: Cell<Operation>,
+
+    value_buffer: Cell<Option<&'static [u8]>>,
+    key_buffer: TakeCell<'static, [u8; 8]>,
+    ret_buffer: TakeCell<'static, [u8]>,
+
+    client: OptionalCell<&'a dyn kv_system::Client<TicKVKeyType>>,
+}
+
+impl<'a, F: Flash> TicKVStore<'a, F> {
+    pub fn new(
+        flash: &'a F,
+        tickfs_read_buf: &'static mut [u8; 512],
+        flash_read_buffer: &'static mut F::Page,
+        region_offset: usize,
+        flash_size: usize,
+    ) -> TicKVStore<'a, F> {
+        let tickv = AsyncTicKV::<TickFSFlastCtrl<F>, 512>::new(
+            TickFSFlastCtrl::new(flash, flash_read_buffer, region_offset),
+            tickfs_read_buf,
+            flash_size,
+        );
+
+        Self {
+            tickv,
+            operation: Cell::new(Operation::None),
+            next_operation: Cell::new(Operation::None),
+            value_buffer: Cell::new(None),
+            key_buffer: TakeCell::empty(),
+            ret_buffer: TakeCell::empty(),
+            client: OptionalCell::empty(),
+        }
+    }
+
+    pub fn initalise(&self) {
+        let _ret = self.tickv.initalise(0x7bc9f7ff4f76f244);
```

Comment:
could this be a labeled const?

Reply:
The function?

Reply:
The function can't be a const, Rust throws some errors

Reply:
Sorry, I just meant that value passed to initialise. At a glance it wasn't clear to me whether that particular value was some random number or some special value

Reply:
Ah, I understand. This value is the key hash of the main key (used to identify if flash is already setup).

I'm hoping that it can be removed and just calculated when a hash algorithm is added (the next step). Otherwise it should be moved to a const, but probably when the hash algorithm is set so then it's more official.

Reply:
Got it, sounds good